Project5_II/python_sql.py

```diff
@@ -0,0 +1,96 @@
+import os
+import gspread
+import logging
+import psycopg2
+from dotenv import load_dotenv
+
+load_dotenv()
+#configure logging library
+logging.basicConfig(filename='errors.log', level=logging.ERROR,
+                    format = '%(levelname)s (%(asctime)s) : %(message)s (%(lineno)d)')
+
+def setup():
+    """authorizes gspread and gets data
+    from spreadsheet"""
+    #get current working directory
+    current_working_directory = os.getcwd()
+    #get environment variable project_json
+    project_json = os.getenv("PROJECT_JSON")
+    path = os.path.join(current_working_directory,project_json)
+
+    #authenticate gspread
+    client = gspread.service_account(filename = path)
+    #connect to online spreadsheet
+    spreadsheet = client.open("Gspread SpreadSheet")
+    #access worksheet in spreadsheet
+    worksheet = spreadsheet.worksheet("Sheet1")
+    #get all values from worksheet
+    github_data = worksheet.get_all_values()
+    #github_data[1][0] = 1
+    
+    return github_data
+
+# Connect to an existing database
+def validate_data(github_data):
+    """function checks that each
+    element in each row is of expected 
+    datatype """
+    data = []
+
+    #loop through data and check if each is of expected data type
+    for num in range(1,len(github_data)):
+        if (isinstance(github_data[num][0], str) and
+            isinstance(github_data[num][1], str) and
+            isinstance(github_data[num][2], str) and
+            isinstance(github_data[num][3], str)):
+            
+            data.append(github_data[num])
+        else:
+            #log error message to error.log file
+            logging.error(f"Invalid Data Type in row")
+            
+    return data
+
+
+def database_connect_execute():
+    """connects to postgres database
+    and executes sql commands to perform 
+    operations"""
+    #get password from environment variable
+    PASSWORD = os.getenv("PASSWORD")
+    #connect to postgres database
+    with psycopg2.connect(host="localhost",dbname="GspreadData_Storage",
```

Comment:
Your db name needs to look better. You can stick with lower case and use a descriptive name - `gspreaddata_storage` doesn't sound good. You should name your db based on the data, use case or task (I mean what is being achieved)